aip/client-libraries/4234.md

```diff
@@ -0,0 +1,102 @@
+---
+id: 4234
+state: draft
+created: 2021-02-10
+---
+
+# Common service client mix-ins
+
+Often, APIs can have common features for administering resources that are
+separated into utility services. These common, utility services are shared among
+the discrete product services. These common services each have a
+centrally-defined surface, but individual instances are hosted alongside each
+product service. Thus, the surfaces of each common service hosted by the product
+service are "mixed in" at runtime.
+
+
+## Guidance
+
+Client libraries **may** provide client methods for the common services declared
+to be mixed-in by a product service. Such methods improve the user experience by
+presenting the mix-in methods from a client configured to communicate with the
+product service that hosts the mix-in service.
+
+Should client libraries support mix-in services, the common services that
+**must** be supported include:
+
+- `google.cloud.location.Locations`
+- `google.iam.v1.IAMPolicy`
+- `google.longrunning.Operations`
+
+**Note:** The list of supported mix-in services and the mix-in services
+themselves change infrequently. Any new common service to be supported should be
+added to the list here and generators should be updated.
+
+To be generated into a client library, a mix-in service **must** be declared
+under [apis] in a [google.api.Service]. Furthermore, it **must** have
+[google.api.http] bindings declared for all of its RPCs in the [http]
+configuration of the same [google.api.Service] (more details on these bindings
+in AIP-127).
+
+
+## Implementing mix-in support
+
+### Generator configuration
+
+Client library generators **must** accept the file path of a
+[google.api.Service] in YAML form. This flag **must** be optional - meaning
```

Comment:
Is there a public description of the file format anywhere?

Reply:
Yep, [here](https://github.com/googleapis/googleapis/blob/8b3d36daaf5561496b7d4075fba4f2c52d18ca1c/google/api/service.proto). That's what [google.api.Service] links to (see bottom of the file).

Reply:
I outlined my concern for this in [this comment](https://github.com/aip-dev/google.aip.dev/pull/694#discussion_r591855538). Resolving this in the interest of having one single thread.

Reply:
It actually works as is, without the trailing `[]`, verified as you suggested :)